crates/re_build_info/src/crate_version.rs

```diff
@@ -1,94 +1,191 @@
+    };
+    ($s:expr, $start:expr, $end:expr) => {
+        __slice($s, Some($start), Some($end))
+    };
+}
+
+const fn equals(a: &[u8], b: &[u8]) -> bool {
+    if a.len() != b.len() {
+        return false;
+    }
+
+    let len = a.len();
+    let mut i = 0;
+    while i < len {
+        if a[i] != b[i] {
+            return false;
+        }
+        i += 1;
+    }
+    true
+}
+
+const fn split_at(s: &[u8], i: usize) -> (&[u8], &[u8]) {
+    (slice!(s, .., i), slice!(s, i, ..))
 }
 
 impl CrateVersion {
     pub const fn new(major: u8, minor: u8, patch: u8) -> Self {
-        assert!(
-            major <= MAX_NUM && minor <= MAX_NUM && patch <= MAX_NUM,
-            "Too large number in version string"
-        );
         Self {
             major,
             minor,
             patch,
-            alpha: None,
-            prerelease: false,
+            meta: None,
         }
     }
 
     /// Whether or not this build is a prerelease (a version ending with +commit suffix)
-    pub fn is_prerelease(&self) -> bool {
-        self.prerelease
+    pub fn is_dev(&self) -> bool {
+        matches!(self.meta, Some(Meta::DevAlpha(..)))
     }
 
-    /// From a compact 32-bit representation crated with [`Self::to_bytes`].
-    pub fn from_bytes([major, minor, patch, suffix_byte]: [u8; 4]) -> Self {
-        let is_alpha = (suffix_byte & IS_ALPHA_BIT) != 0;
-        let is_prerelease = (suffix_byte & IS_PRERELEASE_BIT) != 0;
-        let alpha_version = suffix_byte & !(IS_ALPHA_BIT | IS_PRERELEASE_BIT);
+    pub fn is_alpha(&self) -> bool {
+        matches!(self.meta, Some(Meta::Alpha(..) | Meta::DevAlpha(..)))
+    }
+
+    pub fn is_rc(&self) -> bool {
+        matches!(self.meta, Some(Meta::Rc(..)))
+    }
 
+    /// From a compact 32-bit representation crated with [`Self::to_bytes`].
+    pub fn from_bytes([major, minor, patch, meta]: [u8; 4]) -> Self {
         Self {
             major,
             minor,
             patch,
-            alpha: is_alpha.then_some(alpha_version),
-            prerelease: is_prerelease,
+            meta: Meta::from_byte(meta),
         }
     }
 
     /// A compact 32-bit representation. See also [`Self::from_bytes`].
     pub fn to_bytes(self) -> [u8; 4] {
-        let Self {
-            major,
-            minor,
-            patch,
-            alpha,
-            prerelease,
-        } = self;
-
-        let mut suffix_byte = if let Some(alpha) = alpha {
-            IS_ALPHA_BIT | alpha
-        } else {
-            0
-        };
-
-        suffix_byte |= if prerelease { IS_PRERELEASE_BIT } else { 0 };
-
-        [major, minor, patch, suffix_byte]
+        [
+            self.major,
+            self.minor,
+            self.patch,
+            self.meta.map(Meta::to_byte).unwrap_or_default(),
+        ]
     }
 
     pub fn is_compatible_with(self, other: CrateVersion) -> bool {
-        if self.alpha != other.alpha {
+        if self.meta != other.meta {
```

Comment:
this also makes 0.10.0 incompatible with 0.10.1-rc as well as rc versions among each other all of which might cause the wrong experience during rc testing